fabtests/unit/getinfo_test.c

```diff
@@ -964,6 +964,477 @@ getinfo_test(caps, 4, "Test for capability bit regression",
+		return -FI_EINVAL;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int test_invalid_optional_cap(char *node, char *service, uint64_t flags,
+				     struct fi_info *hints,
+				     struct fi_info **info)
+{
+	uint64_t caps;
+	uint64_t domain_caps;
+	uint64_t tx_caps;
+	uint64_t rx_caps;
+	int ret;
+
+	/* Cache provider supported caps used for validation later. */
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		sprintf(err_buf, "No providers found");
+		return -FI_ENOSYS;
+	}
+
+	caps = (*info)->caps;
+	domain_caps = (*info)->domain_attr->caps;
+	tx_caps = (*info)->tx_attr->caps;
+	rx_caps = (*info)->rx_attr->caps;
+
+	/* FI_VARIABLE_MSG is used for invalidate optional cap. Skip test if
+	 * this cap is supported.
+	 */
+	if (caps & FI_VARIABLE_MSG) {
+		sprintf(err_buf, "FI_VARIABLE_MSG is supported by %s",
+			(*info)->domain_attr->name);
+		return -FI_ENOSYS;
+	}
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	/* Feed in all unsupported capabilities as optional caps. Many of the
+	 * unsupported caps are highly likely to be unset in returned fi_infos
+	 */
+	hints->caps = caps;
+	hints->optional_caps = FI_VARIABLE_MSG;
+	hints->domain_attr->caps = domain_caps;
+	hints->domain_attr->optional_caps = FI_VARIABLE_MSG;
+	hints->tx_attr->caps = tx_caps;
+	hints->tx_attr->optional_caps = FI_VARIABLE_MSG;
+	hints->rx_attr->caps = rx_caps;
+	hints->rx_attr->optional_caps = FI_VARIABLE_MSG;
+
+	/* Verify that invalid optional_caps are not returned in the fi_info. In
+	 * addition, fi_getinfo() should not fail.
+	 */
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_UNIT_STRERR(err_buf,
+			       "fi_getinfo with optional_caps failed to return",
+			       ret);
+		return -FI_EINVAL;
+	}
+
+	ret = test_invalid_cap_info(*info, FI_VARIABLE_MSG, FI_VARIABLE_MSG,
+				    FI_VARIABLE_MSG);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf,
+			       "FI_VARIABLE_MSG incorrectly returned in caps",
+			       ret);
+		return ret;
+	}
+
+	ret = test_valid_optional_cap_info(*info, 0, 0, 0);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "Optional caps incorrectly returned",
+			       ret);
+		return ret;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int test_valid_optimal_optional_cap(char *node, char *service,
+					   uint64_t flags,
+					   struct fi_info *hints,
+					   struct fi_info **info)
+{
+	uint64_t required_caps = FI_MSG | FI_TAGGED;
+	int ret;
+
+	/* Verify that the provider meets the required capabilities of this
+	 * test.
+	 */
+	hints->caps = required_caps;
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		sprintf(err_buf, "No providers found");
+		return -FI_ENOSYS;
+	}
+
+	/* This test is only setup to support selected providers. */
+	if (!strstr(hints->fabric_attr->prov_name, "ofi_rxm")) {
+		sprintf(err_buf, "%s not support for this test",
+			hints->fabric_attr->prov_name);
+		return -FI_ENOSYS;
+	}
```

Comment:
I think we can remove this and test any provider

Reply:
I think that is safe to remove for this test. The subsequent test, test_valid_suboptimal_optional_cap, is more RxM specific since it expects specific handling of FI_ATOMIC optional_cap.